packages/core/src/lib/routes/passkey-options.ts

```diff
@@ -0,0 +1,125 @@
+ */
+async function doRegister(options: InternalOptions<PasskeyProviderType>, email: string): Promise<ResponseInternal<PasskeyOptionsReturn>> {
+  // Get registration options
+  const regOptions = await registrationOptions(options, email)
+
+  // Set the cookie
+  const cookieData: PasskeyOptionsCookieData = {
+    providerAccountId: regOptions.user.id,
+    challenge: regOptions.challenge,
+  }
+  const [cookie] = await signCookie("challenge", cookieData, options)
+
+  // Return the options and set the challenge cookie
+  return {
+    status: 200,
+    body: { options: regOptions, action: "register" },
+    cookies: [cookie],
+  }
+}
+
+/**
+ * Generate passkey authentication options and set the challenge cookie.
+ */
+async function doAuthenticate(options: InternalOptions<PasskeyProviderType>, email?: string): Promise<ResponseInternal<PasskeyOptionsReturn>> {
+  // Get auth options
+  const authOptions = await authenticationOptions(options, email)
+
+  // Set the cookie
+  const cookieData: PasskeyOptionsCookieData = {
+    challenge: authOptions.challenge,
+  }
+  const [cookie] = await signCookie("challenge", cookieData, options)
+
+  // Return the options and set the challenge cookie
+  return {
+    status: 200,
+    body: { options: authOptions, action: "authenticate" },
+    cookies: [cookie],
+  }
+}
+
+/**
+ * Handle passkey options requests by generating authentication or registration options
+ * based on the query parameters and the user's credentials.
+ *
+ * @param request The incoming request.
+ * @param options
+ * @returns A response with the options and a signed challenge cookie.
+ */
+export async function passkeyOptions(
+  options: InternalOptions<PasskeyProviderType>,
+  sessionStore: SessionStore,
+  action?: string,
+  queryEmail?: string
+): Promise<ResponseInternal<PasskeyOptionsReturn> | ResponseInternal<string>> {
+  const { adapter } = options
+
+  let selectedAction: string | undefined = action
+
+  // Get the current session, if it exists
+  // NOTE: this is a bit hacky, but routes.session seems to be
+  // the only place that implements a full session/user check.
+  const { body: currentSession } = await routesSession({
+    options,
+    sessionStore,
+  })
+  const sessionUserEmail = currentSession?.user?.email ?? undefined
+  const loggedIn = !!sessionUserEmail
+
+  const providedQueryEmail = !!queryEmail
+
+  // Find whether a user with queryEmail exists
+  let userWithEmailExists = false
+  if (providedQueryEmail) {
+    const existingUser = await adapter?.getUserByEmail(queryEmail)
+    userWithEmailExists = !!existingUser
+  }
+
+  // This logic is explained in the original PR:
+  // https://github.com/nextauthjs/next-auth/pull/8808
+  switch (selectedAction) {
+    case "authenticate": {
+      if (!loggedIn) {
+        return doAuthenticate(options, queryEmail)
+      }
+
+      break
+    }
+    case "register": {
+      if ((loggedIn !== providedQueryEmail) && !userWithEmailExists) {
```

Comment:
Registering could also happens if the same user want to create another authenticator, right? Then I think we don't need to check for `!userWithEmailExist` 🤔

Reply:
`userWithEmailExists` is true if the value sent via the `email` query param belongs to an existing user.

If a user wants to register a new authenticator, they will make an authenticated request and will omit the `email` param.

We can't accept the `email` query to add an extra authenticator because that would allow anyone to register new authenticators to existing users

Reply:
got it, thanks for the explanation